clippy_lints/src/use_self.rs

```diff
@@ -81,78 +70,189 @@ fn span_use_self_lint(cx: &LateContext<'_, '_>, path: &Path<'_>, last_segment: O
+        _ => sp,
+    }
+}
+
+struct ImplVisitor<'a, 'tcx> {
     cx: &'a LateContext<'a, 'tcx>,
+    item: &'tcx Item<'tcx>,
     self_ty: Ty<'tcx>,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for SemanticUseSelfVisitor<'a, 'tcx> {
+impl<'a, 'tcx> ImplVisitor<'a, 'tcx> {
+    fn check_trait_method_impl_decl(
+        &mut self,
+        impl_item: &ImplItem<'tcx>,
+        impl_decl: &'tcx FnDecl<'tcx>,
+        impl_trait_ref: ty::TraitRef<'tcx>,
+    ) {
+        let tcx = self.cx.tcx;
+        let trait_method = tcx
+            .associated_items(impl_trait_ref.def_id)
+            .find_by_name_and_kind(tcx, impl_item.ident, ty::AssocKind::Fn, impl_trait_ref.def_id)
+            .expect("impl method matches a trait method");
+
+        let trait_method_sig = tcx.fn_sig(trait_method.def_id);
+        let trait_method_sig = tcx.erase_late_bound_regions(&trait_method_sig);
+
+        let output_hir_ty = if let FnRetTy::Return(ty) = &impl_decl.output {
+            Some(&**ty)
+        } else {
+            None
+        };
+
+        // `impl_hir_ty` (of type `hir::Ty`) represents the type written in the signature.
+        // `trait_ty` (of type `ty::Ty`) is the semantic type for the signature in the trait.
+        // We use `impl_hir_ty` to see if the type was written as `Self`,
+        // `hir_ty_to_ty(...)` to check semantic types of paths, and
+        // `trait_ty` to determine which parts of the signature in the trait, mention
+        // the type being implemented verbatim (as opposed to `Self`).
+        for (impl_hir_ty, trait_ty) in impl_decl
+            .inputs
+            .iter()
+            .chain(output_hir_ty)
+            .zip(trait_method_sig.inputs_and_output)
+        {
+            // Check if the input/output type in the trait method specifies the implemented
+            // type verbatim, and only suggest `Self` if that isn't the case.
+            // This avoids suggestions to e.g. replace `Vec<u8>` with `Vec<Self>`,
+            // in an `impl Trait for u8`, when the trait always uses `Vec<u8>`.
+            // See also https://github.com/rust-lang/rust-clippy/issues/2894.
+            let self_ty = impl_trait_ref.self_ty();
+            if !trait_ty.walk().any(|inner| inner == self_ty.into()) {
+                self.visit_ty(&impl_hir_ty);
+            }
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ImplVisitor<'a, 'tcx> {
     type Map = Map<'tcx>;
 
-    fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty<'_>) {
-        if let TyKind::Path(QPath::Resolved(_, path)) = &hir_ty.kind {
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::OnlyBodies(self.cx.tcx.hir())
+    }
+
+    fn visit_ty(&mut self, hir_ty: &'tcx hir::Ty<'tcx>) {
+        if let TyKind::Path(QPath::Resolved(_, path)) = hir_ty.kind {
             match path.res {
                 def::Res::SelfTy(..) => {},
                 _ => {
                     if hir_ty_to_ty(self.cx.tcx, hir_ty) == self.self_ty {
```

Comment:
Commenting out this block makes the other tests pass.

Reply:
I could use a little help on this one, I'm a bit clueless :).

Reply:
So this call results in a `delay_span_bug` here: https://github.com/rust-lang/rust/blob/e91aebc1a3835b9b420da0c021e211175a724b8d/src/librustc_typeck/collect.rs#L309-L312

But this method can only be called, if `hir_ty.kind == hir::TyKind::Infer` ([Code](https://github.com/rust-lang/rust/blob/e91aebc1a3835b9b420da0c021e211175a724b8d/src/librustc_typeck/astconv.rs#L2792-L2798). Since we're checking that the `hir_ty.kind == Path` earlier, this doesn't make any sense.

But it gets weirder. if we insert a `dbg!(hir_ty)` or `println!("{:?}", hir_ty)` it doesn't panic anymore and the `hir_ty.kind` is always printed as `TyKind::Path(Resolved)`... 

In fact, even a `print!("x");` before this line makes the ICE go away. It has to print something though, `print!("");` won't work.

Since println debugging won't help here, I will investigate this further with `gdb`.

Reply:
Two backtraces of the ICEs for the errors 

"bad placeholder type" (`eta.rs`, `question_mark.rs`)

```
error: internal compiler error: src/librustc_typeck/collect.rs:311: bad placeholder type
   --> tests/ui/eta.rs:116:15
    |
116 |         Thunk(Box::new(move || option.take().unwrap()()))
    |               ^^^^^^^^

thread 'rustc' panicked at 'Box<Any>', /home/pkrones/rust/src/libstd/macros.rs:13:23
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::sys_common::backtrace::print
   4: std::panicking::default_hook::{{closure}}
   5: std::panicking::default_hook
   6: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at /home/pkrones/rust/src/liballoc/boxed.rs:1048
   7: clippy_driver::report_clippy_ice
             at src/driver.rs:242
   8: <clippy_driver::ICE_HOOK as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}
             at src/driver.rs:235
   9: std::panicking::rust_panic_with_hook
  10: std::panicking::begin_panic
  11: rustc_errors::HandlerInner::span_bug
  12: rustc_errors::Handler::span_bug
  13: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
  14: rustc_middle::ty::context::tls::with_opt::{{closure}}
  15: rustc_middle::ty::context::tls::with_opt
  16: rustc_middle::util::bug::opt_span_bug_fmt
  17: rustc_middle::util::bug::span_bug_fmt
  18: <rustc_typeck::collect::ItemCtxt as rustc_typeck::astconv::AstConv>::ty_infer
  19: <dyn rustc_typeck::astconv::AstConv>::create_substs_for_ast_path::{{closure}}
  20: <dyn rustc_typeck::astconv::AstConv>::create_substs_for_generic_args
  21: <dyn rustc_typeck::astconv::AstConv>::create_substs_for_ast_path
  22: <dyn rustc_typeck::astconv::AstConv>::ast_path_substs_for_ty
  23: <dyn rustc_typeck::astconv::AstConv>::res_to_ty
  24: <dyn rustc_typeck::astconv::AstConv>::ast_ty_to_ty
  25: rustc_typeck::hir_ty_to_ty
  26: <clippy_lints::use_self::ImplVisitor as rustc_hir::intravisit::Visitor>::visit_ty
             at clippy_lints/src/use_self.rs:166
...
```


"unelided lifetime in signature" (`crashes/ice-4775.rs`, `expect_fun_call.rs`, `unit_arg.rs`)

```
error: internal compiler error: src/librustc_typeck/astconv.rs:182: unelided lifetime in signature                                                                                                                                                --> /home/pkrones/rust/src/libstd/macros.rs:97:28                                                                                                                                                                                                |                                                                                                                                                                                                                                            97 |         $crate::io::_print($crate::format_args_nl!($($arg)*));                                                                                                                                                                                |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                                                                                                               thread 'rustc' panicked at 'Box<Any>', /home/pkrones/rust/src/libstd/macros.rs:13:23                                                                                                                                                            stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::sys_common::backtrace::print
   4: std::panicking::default_hook::{{closure}}
   5: std::panicking::default_hook
   6: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at /home/pkrones/rust/src/liballoc/boxed.rs:1048
   7: clippy_driver::report_clippy_ice
             at src/driver.rs:242
   8: <clippy_driver::ICE_HOOK as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}
             at src/driver.rs:235
   9: std::panicking::rust_panic_with_hook
  10: std::panicking::begin_panic
  11: rustc_errors::HandlerInner::span_bug
  12: rustc_errors::Handler::span_bug
  13: rustc_middle::util::bug::opt_span_bug_fmt::{{closure}}
  14: rustc_middle::ty::context::tls::with_opt::{{closure}}
  15: rustc_middle::ty::context::tls::with_opt
  16: rustc_middle::util::bug::opt_span_bug_fmt
  17: rustc_middle::util::bug::span_bug_fmt
  18: <dyn rustc_typeck::astconv::AstConv>::ast_region_to_region::{{closure}}
  19: <dyn rustc_typeck::astconv::AstConv>::ast_region_to_region
  20: <dyn rustc_typeck::astconv::AstConv>::create_substs_for_generic_args
  21: <dyn rustc_typeck::astconv::AstConv>::create_substs_for_ast_path
  22: <dyn rustc_typeck::astconv::AstConv>::ast_path_substs_for_ty
  23: <dyn rustc_typeck::astconv::AstConv>::res_to_ty
  24: <dyn rustc_typeck::astconv::AstConv>::ast_ty_to_ty
  25: rustc_typeck::hir_ty_to_ty
  26: <clippy_lints::use_self::ImplVisitor as rustc_hir::intravisit::Visitor>::visit_ty
             at clippy_lints/src/use_self.rs:166
...
```

Reply:
@flip1995, is this a misuse of the api or a compiler bug?

Reply:
I'm not sure. The `hir_ty_to_ty` should never panic. At least I don't expect it to panic. 

@eddyb Do you have an idea, why `hir_ty_to_ty` panics on some `hir::Ty`s in the `visit_ty` visitor method?

Reply:
Could this be related to the `QPath::TypeRelative` span bug? See conversation here: https://discordapp.com/channels/442252698964721669/459149871191031810/703737889049280512
Basically for `QPath::TypeRelative(hir_ty, path_segment)` variants, the span of the underlying `hir_ty` incorrectly covers the whole `qpath`. Concretely, in the function call `Box::new(..)`, the function path is a `qpath @ QPath::TypeRelative(hir_ty, segment)`. The span of `qpath` is `Box::new` and the currently incorrect span of the `hir_ty` is also `Box::new`.

Reply:
Ignoring `HirTy`s when the upper node is a `QPath::TypeRelative` in `visit_ty` fixes the failing lint tests (but introduces false negatives).
```rust
                    if let Some(Node::Expr(Expr {
                        kind: ExprKind::Path(QPath::TypeRelative(_, segment)),
                        ..
                    })) = self.cx.tcx.hir().find(self.cx.tcx.hir().get_parent_node(hir_ty.hir_id))
                    {
                        ()
                    } else {
                        if hir_ty_to_ty(self.cx.tcx, hir_ty) == self.self_ty {
                            span_lint(self.cx, hir_ty.span);
                        }
                    }
```

Reply:
It probably isn't connected to the span bug.

How many FNs does this introduce? I would be good with this hack, if only a few FN are introduced.

Reply:
Quite a few. I committed the hack and flagged all FN in the ui test `use_self.rs` file.

I left a commented block explaining the problem and indicating which code to use in order to resolve most false negatives. We could open a follow-up issue to track the unresolved ICEs.

I think we're better off with all those false negatives rather than the false positives of the previous implementation. The lint will catch less cases, but will be more accurate.

Reply:
> We could open a follow-up issue to track the unresolved ICEs.

We can't merge a PR that introduces ICEs. Did you mean for the FPs?

> I think we're better off with those false negatives rather than the false positives of the previous implementation. The lint will catch less cases, but will be more accurate.

Agreed ✔️

Reply:
> We can't merge a PR that introduces ICEs. Did you mean for the FPs?
In the current patch state that passes all tests, a block of the implementation is commented out to avoid ICEs. We could create a follow-up ticket to investigate this and try to make that part work without errors. That would solve most FNs.

Reply:
Oh now the dogfood_clippy test throws errors :disappointed:.

Reply:
That sadly tells us, that it is not possible to fix the ICEs by excluding specific `TyKind`s. I kind of expected that, since `TyKind` is a recursive enum, so just checking the top level type doesn't really help, since some subtype could still be problematic.

I will look in to this closer, when I have some time. I hope I get to it rather sooner than later.